### PR TITLE
PUBDEV-8287: TreeSHAP reweighting API

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

File: h2o-py/h2o/model/model_base.py [hunk 1]
```diff
@@ -516,6 +516,19 @@ def h(self, frame, variables):
             return self._h(frame=frame, variables=variables)
         print("No calculation available for this model")
 
+    def update_tree_weights(self, frame, weights_column):
+        """
+        Re-calculates tree-node weights based on provided dataset. Modifying node weights will affect how
```
Comment (michalkurka) [inline h2o-py/h2o/model/model_base.py:521]: @meganjkurka does this description look okay to you?

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java [hunk 2]
```diff
@@ -488,6 +492,8 @@ protected Frame execute(Frame adaptFrm, String[] names, Key<Frame> destKey) {
     private transient int[/*treeId*/][/*classId*/] _maxNodeIds;
     // OUT
     private double[/*treeId*/][/*classId*/][/*leafNodeId*/] _leafNodeWeights;
+    private int[] _warn_trees;
```
Comment (valenad1) [inline h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:495]: I would use camelCase to be in consistent with previous variables 😇

File: h2o-core/src/main/java/hex/Model.java [hunk 3]
```diff
@@ -117,7 +117,15 @@ public void setInputParms(P _input_parms) {
   }
 
   public interface UpdateAuxTreeWeights {
-    void updateAuxTreeWeights(Frame frame, String weightsColumn);
+    UpdateAuxTreeWeightsReport updateAuxTreeWeights(Frame frame, String weightsColumn);
+
+    class UpdateAuxTreeWeightsReport {
+      public int[] _warn_trees;
```
Comment (valenad1) [inline h2o-core/src/main/java/hex/Model.java:123]: Also camelCase

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java [hunk 4]
```diff
@@ -723,6 +723,22 @@ public void updateAuxTreeWeights(Frame frame, String weightsColumn) {
...
+    UpdateAuxTreeWeightsReport report = new UpdateAuxTreeWeightsReport();
+    report._warn_classes = new int[0];
+    report._warn_trees = new int[0];
+    for (int treeId = 0; treeId < nodeWeights.length; treeId++) {
+      if (nodeWeights[treeId] == null)
+        continue;
+      for (double w : nodeWeights[treeId]) {
+        if (w == 0) {
+          report._warn_trees = ArrayUtils.append(report._warn_trees, treeId);
+          report._warn_classes = ArrayUtils.append(report._warn_classes, 0);
```
Comment (valenad1) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java:736]: Why we are appending 0?

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java [hunk 5]
```diff
@@ -564,6 +570,8 @@ public void reduce(UpdateAuxTreeWeightsTask mrt) {
 
     @Override
     protected void postGlobal() {
+      _warn_trees = new int[0];
```
Comment (valenad1) [inline h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:573]: Also camelCase

**Review by valenad1 — APPROVED** [review]
Thank you! I have found only minors.

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java (hunk 2 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:495]: yes, here it is inconsistent

File: h2o-core/src/main/java/hex/Model.java (hunk 3 above)
Comment (michalkurka) [reply h2o-core/src/main/java/hex/Model.java:123]: this is the better way for public-facing classes

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java (hunk 5 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:573]: yep, thank you

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java (hunk 4 above)
Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java:736]: class 0, multinomial is not supported but could be added in the future without API change

**Pull request merged**